### ISSUE TITLE
G: Writing our own code for Lasso regression

### ISSUE THREAD

**Issue by SigurdRiseth** [SigurdRiseth/FYS-STK4155-P1#10]
Lasso regression (Section 3.9 of the lecture notes, and the lectures of weeks 35 and 36) represents our first encounter with a machine learning method which cannot be solved through analytical expressions (as in OLS and Ridge regression).  Use the gradient descent methods you developed in parts e) and f) to solve the Lasso optimization problem.

The $\ell_1$ penalty $\lambda\|\boldsymbol{\theta}\|_1$ is not differentiable at $\theta_j=0$.  Discuss how you could handle this: what does your automatic differentiation tool return for the derivative of $|\theta|$ at zero, and is that a valid subgradient? Based on the gradient, write your own code for Lasso regression.

You can compare your results with the functionalities of **Scikit-Learn**; be careful with the different conventions for the penalty parameter (`alpha`) and the normalization of the cost function discussed in the week-35 and week-36 Tuesday sessions.

Discuss (critically) your results for the Runge function from OLS, Ridge and Lasso regression using the various gradient descent approaches.
